### PR TITLE
Allow genbranch to generate the same commits over multiple invocations

### PR DESCRIPTION
**Pull request by lucasdemarchi**
This changes the arguments to `git am` so the commits generated by applying the patches have the same SHAs if the baseline is the same and the patches are the same.

If the baseline is the same between 2 invocations of `git pile genbranch` the other 2 variables that are accounted for during the hash calculation are: the committer date, committer name and commiter's email. Now we can set these to fixed values: name and email we read from config; date we consider the same as author's.

This allows the generated branch to have the same git objects after multiple invocations, even if done by different developers.

### CODE REVIEW

**Review by ashutoshx — APPROVED** [review]
After googling around a bit I think I understand how git generates SHA's, e.g. link below, so I think I know what's happening.

https://stackoverflow.com/questions/34764195/how-does-git-create-unique-commit-hashes-mainly-the-first-few-characters

So for all commits generated by genbranch we will replace the commiter date by the author date and committer SHA with the author SHA, which seems fine. So I am trying to think if we lose anything in this process.

Basically we will lose when genbranch was done and who did the genbranch (the committer info). If we want this the best we have is the info from the pile branch. So we have that so all is not lost.

So overall I think should be ok, just maybe need to get used to the new way things are going to be.

File: git_pile/git_pile.py
Comment (ashutoshx) [inline git_pile/git_pile.py:1151]: My guess about what is happening here is that we are setting committer name and email to "None" which will induce git to use the author name and email, correct, when doing 'git am' etc.? Is there any particular reason why we don't set these to None directly rather than set genbranch_user_name and email to None? Is it just to show the genbranch user name/email is being eliminated?

**Review by ashutoshx — APPROVED** [review]
Another idea, if it's worth it, would be to retain author name/date only when we know that all SHA's are the same, if we know that some commits have changed so the SHA's will be different anyway we might as well as retain the committer name/date rather than the author name/date. But I don't know if git pile can do that (track parent commits etc.). So maybe leave as is. The series is (sort of):

Reviewed-by: Ashutosh Dixit <ashutosh.dixit@intel.com>

**Comment by lucasdemarchi** [comment]
> So for all commits generated by genbranch we will replace the commiter date by the author date and committer name with the author name, which seems fine. So I am trying to think if we lose anything in this process.

- Almost that. committer date will be author date
- Committer email/name will be whatever is set in the config

The idea is to leave the committer as something generic that can be shared by the developers. Example: "Awesome Bot <awesome.bot@example.com>".

Note that today we can already change the commiter email/name by simply exporting the correct env vars:

```bash
export GIT_COMMITER_NAME="Awesome Bot" GIT_COMMITER_EMAIL=awesome.bot@example.com
git pile genbranch
```

However, some people set the commiter name/email via env var and it would be troublesome to set it only during genbranch. This is the reason I thought it would be neater to have this info in the pile's config section.

**Comment by lucasdemarchi** [comment]
> Basically we will lose when genbranch was done and who did the genbranch (the committer info). If we want this the best we have is the info from the pile branch. So we have that so all is not lost.

Yeah, I think the individual commiter date is the only thing that is lost. Unless you are trying to figure out how long it took to generate the branch by comparing the first and last commits, you'd be better off just using the committer date from the pile branch, that is not affected by this change.

**Comment by lucasdemarchi** [comment]
> Another idea, if it's worth it, would be to retain author name/date only when we know that all SHA's are the same, if we know that some commits have changed so the SHA's will be different anyway we might as well as retain the committer name/date rather than the author name/date. But I don't know if git pile can do that (track parent commits etc.). So maybe leave as is. The series is (sort of):

Note that is the committer name/email that changes. The author's email/name remain the same. That approach is not possible since the goal here is that even if genbranch is called by different people, on different machine, the end result is the same and the commit hashes are the same as well. This has lots of benefits:

- Reduced dangling git objects in the repository, so it takes longer to require a `git gc`
- The commit hashs as reported by git range-diff in a cover letter, are more likely to be found on other developers' machines.
- CI can genbranch and easily check if what is pushed to the result-branch really matches the pile branch

File: git_pile/git_pile.py
Comment (lucasdemarchi) [reply git_pile/git_pile.py:1151]: This is an optional feature. If genbranch_user_name and genbranch_user_email are not set in the configuration file, we will not set the environment variables. In this case git will use its normal logic for defining the committer (most likely from whatever is configure as user.email/user.name in the config).

Comment (ashutoshx) [reply git_pile/git_pile.py:1151]: OK, since committer info has to be the same in order for git to generate the same SHA's, it seems the only way this can work is if everyone sets the same value for committer name and email in their config files, correct? There is a committer-date-is-author-date but no committer-info-is-author-info.

Anyway, if this is what we have to do that's fine. Please go ahead and merge this when you're ready. Thanks!

Comment (lucasdemarchi) [reply git_pile/git_pile.py:1151]: Yes, there isn't a  `--committer-info-is-author-info` in git am. In future if git has such a flag we can reserve a word like "author" for this setting to start using it.

For sharing the same config among users, just sharing a git config snippet to be included by their config would be enough.

**Pull request merged**
